src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,152 @@
+			final TPFResponse res = FederationAccessUtils.performRequest(fedAccessMgr, new TPFRequestImpl((TriplePattern) reqTranslation), fm);
+			final List<Triple> resTranslation = new ArrayList<>();
+			for(final Triple i : res.getPayload()) {
+				BindingBuilder bb = BindingBuilder.create();
+				final Var s = Var.alloc("s");
+				final Var p = Var.alloc("p");
+				final Var o = Var.alloc("o");
+				bb.add(s, i.asJenaTriple().getSubject());
+				bb.add(p, i.asJenaTriple().getPredicate());
+				bb.add(o, i.asJenaTriple().getObject());
+				SolutionMapping sm = new SolutionMappingImpl(bb.build());
+				for (final SolutionMapping j : fm.getVocabularyMapping().translateSolutionMapping(sm)) {
+					Binding b = j.asJenaBinding();
+					resTranslation.add(new TripleImpl(b.get(s), b.get(p), b.get(o)));
+				}
+			}
+			final List<Triple> allTriples = new ArrayList<>();
+			res.getMetadata().forEach(allTriples::add);
+			return new TPFResponseImpl(resTranslation, allTriples, res.getNextPageURL(), fm, req, res.getRequestStartTime());
+		}
+		
+		else if(reqTranslation instanceof SPARQLUnionPattern) {
+			final Set<Triple> resTranslation = new HashSet<>();
+			for (final SPARQLGraphPattern i : ((SPARQLUnionPattern) reqTranslation).getSubPatterns()) {
+				if (i instanceof TriplePattern) {
+					final TPFResponse res =FederationAccessUtils.performRequest(fedAccessMgr, new TPFRequestImpl((TriplePattern) i), fm);
+					for(final Triple j : res.getPayload()) {
+						BindingBuilder bb = BindingBuilder.create();
+						final Var s = Var.alloc("s");
+						final Var p = Var.alloc("p");
+						final Var o = Var.alloc("o");
+						bb.add(s, j.asJenaTriple().getSubject());
+						bb.add(p, j.asJenaTriple().getPredicate());
+						bb.add(o, j.asJenaTriple().getObject());
+						SolutionMapping sm = new SolutionMappingImpl(bb.build());
+						for (final SolutionMapping k : fm.getVocabularyMapping().translateSolutionMapping(sm)) {
+							Binding b = k.asJenaBinding();
+							resTranslation.add(new TripleImpl(b.get(s), b.get(p), b.get(o)));
+						}
+					}
+				}
+				//TODO: Add all other option of instance
+				else {
+					throw new FederationAccessException(i.toString(), req, fm);
+				}
+			}
+			final List<Triple> resTriples = new ArrayList<>();
+			resTriples.addAll(resTranslation);
+			return new TPFResponseImpl(resTriples ,resTriples, null, fm, req, null);
+		}
+		
+		//Same for GroupPattern
+		else if(reqTranslation instanceof SPARQLGroupPattern) {
+			for(final SPARQLGraphPattern i : ((SPARQLGroupPattern) reqTranslation).getSubPatterns()) {
+				if (i instanceof TriplePattern) {
+					FederationAccessUtils.performRequest(fedAccessMgr, new TPFRequestImpl((TriplePattern) i), fm);
+					//Translate solutionMapping
+				}
+				//TODO: Add all other option of instance
+				else {
+					throw new FederationAccessException(i.toString(), req, fm);
+				}
+			}
+			//Return intersection of all results
+		}
+		
+		else if(reqTranslation instanceof BGP) {
```

Comment:
Here, I'm not sure if we defined how the mapping-based translation works over a BGP. Anyhow, the result of translating a BGP should be a join between all the translation results of each triple pattern in the BGP.